flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContext.java

```diff
@@ -257,9 +257,23 @@ public void signalNoMoreSplits(int subtask) {
         notifier.notifyReadyAsync(callable, handler);
     }
 
+    /** {@inheritDoc} If the runnable throws an Exception, the corresponding job is failed. */
     @Override
     public void runInCoordinatorThread(Runnable runnable) {
-        coordinatorExecutor.execute(runnable);
+        coordinatorExecutor.execute(wrap(runnable));
+    }
+
+    private Runnable wrap(final Runnable runnable) {
+        return () -> {
+            try {
+                runnable.run();
+            } catch (final Throwable t) {
+                // when using a SheduledThreadPool, uncaught exception handler catches only
+                // exceptions thrown by the threadPool, so manually call it when the exception is
+                // thrown by the runnable
+                coordinatorThreadFactory.uncaughtException(Thread.currentThread(), t);
```

Comment:
I don't think I understand this. The provided `Runnable` should be always executed by the thread from the pool. Any uncaught exception from the `Runnable` should be passed to the uncaught exception handler of the thread pool.

Why do we need this safeguard? Is there anything I'm missing?

Reply:
See the history in the PR. The pb is that now that coordinatorExecutor is a ScheduledExecutorService, UncaughtExceptionHandler is useful only for exception raised by the threadPool as the JVM now uses a ScheduledFutureTask that catches exceptions in its run method. So the exception handler is no more for the runnable exceptions but for exceptions raised by the pool itself. It is also called in a manual try/catch in the runnable execution in `wrap()` (that Till recommended) to workaround the pb with ScheduledFutureTask explained above.

Reply:
👍 Makes sense.

I'm wondering why are we using the ScheduledExecutor here at first place 🤔 

```
        final Executor errorHandlingCoordinatorExecutor =
                (runnable) ->
                        coordinatorExecutor.execute(
                                new ThrowableCatchingRunnable(
                                        this::handleUncaughtExceptionFromAsyncCall, runnable));

        this.notifier = new ExecutorNotifier(workerExecutor, errorHandlingCoordinatorExecutor);
```

We're effectively transforming it into the `Executor` anyway. Would simply switching from `Executors.newScheduledThreadPool` to `Executors.newSingleThreadExecutor` do the trick?

Reply:
Yes I wondered also. But I'm just reacting to a change on master during the review (no luck). In the first version I indeed used `Executors.newSingleThreadExecutor` it worked well but it no longer complies to the signature of the context method.

Reply:
The interface was changed due to [FLIP-182](https://cwiki.apache.org/confluence/display/FLINK/FLIP-182%3A+Support+watermark+alignment+of+FLIP-27+Sources), which schedules a watermark synchronization runnable in the coordinator executor. And the `scheduleAtFixedRate()` method resides in `ScheduledExecutorService`.

Reply:
OK, I see that now, I've missed the `org.apache.flink.runtime.source.coordinator.SourceCoordinatorContext#getCoordinatorExecutor` 🤦 

Thank you both 👍